sw/device/lib/otbn.h

```diff
@@ -0,0 +1,248 @@
+  /**
+   * No application is loaded.
+   */
+  kOtbnAppNotLoaded = 4,
+  /**
+   * The execution of the application on OTBN failed.
+   *
+   * More specific error information can be obtained with
+   * `dif_otbn_get_err_code()`.
+   */
+  kOtbnExecutionFailed = 5,
+} otbn_result_t;
+
+/**
+ * OTBN context structure.
+ *
+ * Use `otbn_init()` to initialize.
+ */
+typedef struct otbn {
+  /**
+   * The OTBN DIF to access the hardware.
+   */
+  dif_otbn_t dif;
+
+  /**
+   * The application loaded or to be loaded into OTBN.
+   */
+  const otbn_app_t *app;
+
+  /**
+   * Is the application loaded into OTBN?
+   */
+  bool app_is_loaded;
+} otbn_t;
+
+/**
+ * Make an embedded OTBN application image available for use.
+ *
+ * Make symbols available that indicate the start and the end of instruction
+ * and data memory regions, as they are stored in the device memory.
+ *
+ * @param app_name name of the application to load, which is typically the
+ *                 name of the main (assembly) source file.
+ */
+#define OTBN_USE_APP(app_name)                            \
+  extern const char _otbn_app_##app_name##__imem_start[]; \
+  extern const char _otbn_app_##app_name##__imem_end[];   \
+  extern const char _otbn_app_##app_name##__dmem_start[]; \
+  extern const char _otbn_app_##app_name##__dmem_end[];
+
+/**
+ * Initialize the OTBN application information structure
+ *
+ * After making all required symbols from the application image available
+ * through `OTBN_USE_APP()`, use this macro to initialize an `otbn_app_t`
+ * struct with those symbols.
+ *
+ * @param app_name name of the application to load
+ * @see OTBN_USE_APP()
+ */
+#define OTBN_APP_T_INIT(app_name)                     \
+  {                                                   \
+    .imem_start = _otbn_app_##app_name##__imem_start, \
+    .imem_end = _otbn_app_##app_name##__imem_end,     \
+    .dmem_start = _otbn_app_##app_name##__dmem_start, \
+    .dmem_end = _otbn_app_##app_name##__dmem_end,     \
+  }
+
+/**
+ * Make a symbol in the OTBN application image available.
+ *
+ * Symbols are typically function or data pointers, i.e. labels in assembly
+ * code.
+ *
+ * @param app_name    name of the application the function is contained in
+ * @param symbol_name name of the symbol (function, label)
+ */
+#define OTBN_USE_PTR(app_name, symbol_name) \
+  extern const char _otbn_app_##app_name##_##symbol_name[];
+
+/**
+ * Initialize an `otbn_ptr_t` structure.
+ */
+#define OTBN_PTR_T_INIT(app_name, symbol_name) \
+  { .ptr_in_cpu_addr_space = _otbn_app_##app_name##_##symbol_name, }
+
+/**
+ * Initialize the OTBN context structure.
+ *
+ * @param ctx the context object
+ * @param dif_config the OTBN DIF configuration (passed on to the DIF)
+ * @return #kOtbnOk if the operation was successful,
+ *         #kOtbnBadArg if a precondition check failed,
+ *         #kOtbnError otherwise
+ */
+otbn_result_t otbn_init(otbn_t *ctx, const dif_otbn_config_t *dif_config);
```

Comment:
Configs are usually taken by value, no?

Reply:
The OTBN DIF takes a pointer currently (it's not yet redone to follow the new parameter/config split). I'd leave it as pointer here until the DIF is redone.